activerecord/lib/active_record/insert_all.rb

```diff
@@ -8,13 +8,21 @@ class InsertAll # :nodoc:
     attr_reader :on_duplicate, :update_only, :returning, :unique_by, :update_sql
 
     def initialize(model, inserts, on_duplicate:, update_only: nil, returning: nil, unique_by: nil, record_timestamps: nil)
-      @model, @connection, @inserts = model, model.connection, inserts.map(&:stringify_keys)
+      @model, @connection = model, model.connection
       @on_duplicate, @update_only, @returning, @unique_by = on_duplicate, update_only, returning, unique_by
       @record_timestamps = record_timestamps.nil? ? model.record_timestamps : record_timestamps
 
       disallow_raw_sql!(on_duplicate)
       disallow_raw_sql!(returning)
 
+      if inserts.first.is_a?(ActiveRecord::Base)
+        inserts = inserts.map(&:attributes)
+        delete_timestamps_from_attributes!(inserts)
+        delete_virtual_columns_from_attributes!(inserts)
+        connection.assign_next_value_if_serial_column_is_nil!(model.columns, model.sequence_name, inserts)
```

Comment:
This is very similar to what [`attributes_for_create`](https://github.com/rails/rails/blob/e446ef82c8174a4bd6c98f670912bb395b3091d6/activerecord/lib/active_record/attribute_methods.rb#L409-L415) does. 

Maybe check out what [`_create_record`](https://github.com/rails/rails/blob/e446ef82c8174a4bd6c98f670912bb395b3091d6/activerecord/lib/active_record/persistence.rb#L1238-L1258) does and use a similar approach?

Reply:
Thank you for your comment.
I have looked into the behavior of the attributes_for_create method.
Although attributes_for_create method [removes the id from attribute_names when the id column (primary key) is nil](https://github.com/rails/rails/blob/e446ef82c8174a4bd6c98f670912bb395b3091d6/activerecord/lib/active_record/attribute_methods.rb#LL412C12-L412C42), unfortunately I don't think this method can be used in this PR.
Because if you want to upsert_all new records and existing records at the same time, as shown below, you cannot remove the id.
```
book1 = Book.create!(name: "Detective Stories 1", author_id: 1)
book2 = Book.new(name: "Detective Stories 2", author_id: 1)
Book.upsert_all([
  book1,
  book2,
])
```